tests/templates/test_subroutines/test_prepselprep.py

```diff
@@ -0,0 +1,353 @@
+    qml.PrepSelPrep(lcu, control)
+    return qml.state()
+
+
+class TestPrepSelPrep:
+    """Test the correctness of the decomposition"""
+
+    # Use these LCUs as test input
+    lcu1 = qml.dot([0.25, 0.75], [qml.Z(2), qml.X(1) @ qml.X(2)])
+    lcu2 = qml.dot([1 / 2, 1 / 2], [qml.Identity(0), qml.PauliZ(0)])
+
+    a = 0.25
+    b = 0.75
+    A = np.array([[a, 0, 0, b], [0, -a, b, 0], [0, b, a, 0], [b, 0, 0, -a]])
+    decomp = qml.pauli_decompose(A)
+    coeffs, unitaries = decomp.terms()
+    unitaries = [qml.map_wires(op, {0: 1, 1: 2}) for op in unitaries]
+    lcu3 = qml.dot(coeffs, unitaries)
+
+    lcu4 = qml.dot([-0.25, -0.75], [qml.Z(2), qml.X(1) @ qml.X(2)])
+    lcu5 = qml.dot([1 + 0.25j, 0 - 0.75j], [qml.Z(3), qml.X(2) @ qml.X(3)])
+    lcu6 = qml.dot([0.5, -0.5], [qml.Z(1), qml.X(1)])
+    lcu7 = qml.dot([0.5, -0.5, 0 + 0.5j], [qml.Z(2), qml.X(2), qml.X(2)])
+    lcu8 = qml.dot([0.5, 0.5j], [qml.X(2), qml.Z(2)])
+
+    dev = qml.device("default.qubit")
+    manual = qml.QNode(manual_circuit, dev)
+    prepselprep = qml.QNode(prepselprep_circuit, dev)
+
+    @pytest.mark.parametrize(
+        ("lcu", "control", "produced_matrix", "expected_matrix"),
+        [
+            (
+                lcu1,
+                0,
+                qml.matrix(prepselprep, wire_order=[0, 1, 2]),
+                qml.matrix(manual, wire_order=[0, 1, 2]),
+            ),
+            (
+                lcu2,
+                "ancilla",
+                qml.matrix(prepselprep, wire_order=["ancilla", 0]),
+                qml.matrix(manual, wire_order=["ancilla", 0]),
+            ),
+            (
+                lcu3,
+                [0],
+                qml.matrix(prepselprep, wire_order=[0, 1, 2]),
+                qml.matrix(manual, wire_order=[0, 1, 2]),
+            ),
+            (
+                lcu4,
+                [0],
+                qml.matrix(prepselprep, wire_order=[0, 1, 2]),
+                qml.matrix(manual, wire_order=[0, 1, 2]),
+            ),
+            (
+                lcu5,
+                [0, 1],
+                qml.matrix(prepselprep, wire_order=[0, 1, 2, 3]),
+                qml.matrix(manual, wire_order=[0, 1, 2, 3]),
+            ),
+            (
+                lcu6,
+                [0],
+                qml.matrix(prepselprep, wire_order=[0, 1, 2]),
+                qml.matrix(manual, wire_order=[0, 1, 2]),
+            ),
+            (
+                lcu7,
+                [0, 1],
+                qml.matrix(prepselprep, wire_order=[0, 1, 2]),
+                qml.matrix(manual, wire_order=[0, 1, 2]),
+            ),
+            (
+                lcu8,
+                [0, 1],
+                qml.matrix(prepselprep, wire_order=[0, 1, 2]),
+                qml.matrix(manual, wire_order=[0, 1, 2]),
+            ),
+        ],
+    )
+    def test_decomposition(self, lcu, control, produced_matrix, expected_matrix):
+        """Test that the template produces the corrent decomposition"""
```

Comment:
It does not check the decomposition, it checks the matrix with a manually constructed circuit.

Reply:
Thanks for pointing this out. I named this test early on before I understood decompositions. I will pick a new name.